File: OCTOBER-2025/03-10-2025/p407.py
class Solution:
    def trapRainWater(self, heightMap: List[List[int]]) -> int:
        rows, cols = len(heightMap), len(heightMap[0])
        minheap = []
        for r in range(rows):
            for c in range(cols):
                if r in [0, rows-1] or c in [0, cols-1]: #border cells
                    heappush(minheap, (heightMap[r][c], r, c))
                    heightMap[r][c] = -1
        res = 0
        max_h = -1
        while minheap:
            h, r, c = heappop(minheap)
            max_h = max(max_h, h)
            res += max_h - h
            dirs = [[r+1, c], [r-1, c], [r, c+1], [r, c-1]]
            for dr, dc in dirs:
                if (dr < 0 or dr >= rows or dc < 0 or dc >= cols or heightMap[dr][dc] == -1):
                    continue
                heappush(minheap, (heightMap[dr][dc], dr, dc))
                heightMap[dr][dc] = -1

        return res
                
        

File: OCTOBER-2025/03-10-2025/p407.java
class Solution {
    public int trapRainWater(int[][] heightMap) {
        var m = heightMap.length;
        var n = heightMap[0].length;
        var volume = new int[m][n];
        for (var i = 0; i < m; i++)
            System.arraycopy(heightMap[i], 0, volume[i], 0, n);

        var upt = true;
        var init = true;
        while (upt) {
            upt = false;
            for (var i = 1; i < m - 1; i++)
                for (var j = 1; j < n - 1; j++) {
                    var val = Math.max(heightMap[i][j], Math.min(volume[i - 1][j], volume[i][j - 1]));
                    if (init || volume[i][j] > val) {
                        volume[i][j] = val;
                        upt = true;
                    }
                }

            init = false;
            for (var i = m - 2; i >= 1; i--)
                for (var j = n - 2; j >= 1; j--) {
                    var val = Math.max(heightMap[i][j], Math.min(volume[i + 1][j], volume[i][j + 1]));
                    if (volume[i][j] > val) {
                        volume[i][j] = val;
                        upt = true;
                    }
                }
        }
        var res = 0;
        for (var i = 1; i < m - 1; i++)
            for (var j = 1; j < n - 1; j++) {
                var volVal = volume[i][j];
                var heightVal = heightMap[i][j];
                if (volVal > heightVal)
                    res += volVal - heightVal;
            }
        return res;
    }
}